presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java

```diff
@@ -108,30 +102,48 @@ public boolean checkFullAndUpdateMemory()
         }
     }
 
-    public Iterator<Page> buildResult()
+    public long getSizeInMemory()
     {
-        List<Type> types = new ArrayList<>(groupByHash.getTypes());
+        long sizeInMemory = groupByHash.getEstimatedSize();
         for (Aggregator aggregator : aggregators) {
-            types.add(aggregator.getType());
+            sizeInMemory += aggregator.getEstimatedSize();
+        }
+        sizeInMemory -= operatorContext.getOperatorPreAllocatedMemory().toBytes();
+        if (sizeInMemory < 0) {
+            sizeInMemory = 0;
         }
+        return sizeInMemory;
+    }
+
+    public long getGroupCount()
+    {
+        return groupByHash.getGroupCount();
+    }
+
+    @Override
+    public Iterator<Page> buildResult()
+    {
+        return buildResult(consecutiveGroupIds());
+    }
 
-        final PageBuilder pageBuilder = new PageBuilder(types);
+    private Iterator<Page> buildResult(Iterator<Integer> groupIds)
```

Comment:
For spilling there will be added version that builds the result in order sorted by grouping key (for disk based merge-join).